Cmake build system breaks with symbolic links

If the `build` directory is a symbolic link, or when the entire source directory is symlinked (and we build from there), the current cmake build runs into trouble finding its own generated files.

To reproduce:
```
git clone --depth=1 https://github.com/bitcoin/bitcoin.git
mkdir build
cd bitcoin/
ln -s ../build .
cmake -Bbuild -DBUILD_GUI=ON
cd build
make
```

```
[ 37%] Building CXX object src/CMakeFiles/bitcoin_node.dir/validation.cpp.o
In file included from /data/src/tmp/bitcoin/build/src/qt/bitcoinqt_autogen/mocs_compilation.cpp:2:
/data/src/tmp/bitcoin/build/src/qt/bitcoinqt_autogen/EWIEGA46WW/moc_addressbookpage.cpp:10:10: fatal error: ../../../../../src/qt/addressbookpage.h: No such file or directory
   10 | #include "../../../../../src/qt/addressbookpage.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [src/qt/CMakeFiles/bitcoinqt.dir/build.make:784: src/qt/CMakeFiles/bitcoinqt.dir/bitcoinqt_autogen/mocs_compilation.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs...
```

This affects the build with Qt as well as the build with multiprocess, both which have generated files.

With autotools this wasn't an issue, so i think it can be considered a regression. But not a high-priority one as it can be worked around by specifying the full paths instead.

I was able to reproduce this on Ubuntu 23.10 on master @ 7590e93 with the repro steps in OP:

<details>
<summary>Details</summary>

```
[ 66%] Automatic RCC for bitcoin_locale.qrc
[ 67%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoinaddressvalidator.cpp.o
[ 67%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bantablemodel.cpp.o
[ 67%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoinqt_autogen/mocs_compilation.cpp.o
[ 67%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoingui.cpp.o
[ 67%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoinamountfield.cpp.o
[ 67%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoin.cpp.o
[ 67%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/bitcoinunits.cpp.o
[ 68%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/clientmodel.cpp.o
[ 68%] Building CXX object src/qt/CMakeFiles/bitcoinqt.dir/csvmodelwriter.cpp.o
In file included from /tmp/31145/bitcoin/build/src/qt/bitcoinqt_autogen/mocs_compilation.cpp:2:
/tmp/31145/bitcoin/build/src/qt/bitcoinqt_autogen/EWIEGA46WW/moc_addressbookpage.cpp:10:10: fatal error: '../../../../../src/qt/addressbookpage.h' file not found
   10 | #include "../../../../../src/qt/addressbookpage.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [src/qt/CMakeFiles/bitcoinqt.dir/build.make:786: src/qt/CMakeFiles/bitcoinqt.dir/bitcoinqt_autogen/mocs_compilation.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
[ 69%] Linking CXX executable bitcoin-wallet
[ 69%] Built target bitcoin-wallet
make[1]: *** [CMakeFiles/Makefile2:1432: src/qt/CMakeFiles/bitcoinqt.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 70%] Linking CXX static library libbitcoin_node.a
[ 70%] Built target bitcoin_node
make: *** [Makefile:146: all] Error 2
```

</details>

However, if the symlink is constructed using an asbolute path then it works correctly. E.g. for me:

```bash
mkdir -p /tmp/31145/build && cd /tmp/31145
git clone --depth=1 https://github.com/bitcoin/bitcoin.git
cd bitcoin
ln -s /tmp/31145/build .
cmake -Bbuild -DBUILD_GUI=ON
cd build
make
```

So the build system seems to be getting tripped up on the relativity of the symlink.

--- a/src/qt/CMakeLists.txt
+++ b/src/qt/CMakeLists.txt
@@ -35,6 +35,7 @@ endfunction()
 #  - https://doc.qt.io/qt-5/cmake-manual.html
 
 set(CMAKE_AUTOMOC ON)
+set(CMAKE_AUTOMOC_MOC_OPTIONS "-p${CMAKE_CURRENT_SOURCE_DIR}")
 set(CMAKE_AUTORCC ON)
 set(CMAKE_AUTOUIC ON)
 set(CMAKE_AUTOUIC_SEARCH_PATHS forms)

--- a/src/qt/test/CMakeLists.txt
+++ b/src/qt/test/CMakeLists.txt
@@ -2,6 +2,8 @@
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or https://opensource.org/license/mit/.
 
+set(CMAKE_AUTOMOC_MOC_OPTIONS "-p${CMAKE_CURRENT_SOURCE_DIR}")
+
 add_executable(test_bitcoin-qt
   apptests.cpp
   optiontests.cpp

